src/Symfony/Component/HttpFoundation/Request.php

```diff
@@ -1836,9 +1836,16 @@ protected function prepareRequestUri()
             $this->server->remove('IIS_WasUrlRewritten');
         } elseif ($this->server->has('REQUEST_URI')) {
             $requestUri = $this->server->get('REQUEST_URI');
-
-            // HTTP proxy reqs setup request URI with scheme and host [and port] + the URL path, only use URL path
-            $uriComponents = parse_url($requestUri);
+            $uriComponents = array();
+
+            // HTTP proxy reqs setup request URI with scheme and host [and port] + the URL path,
+            // only use URL path.
+            if ('' !== $requestUri && '/' !== $requestUri[0]) {
+                $uriComponents = parse_url($requestUri);
+            } elseif (false !== $pos = strpos($requestUri, '#')) {
+                // Remove the fragment.
+                $requestUri = substr($requestUri, 0, $pos);
+            }
```

Comment:
In order to remove extra complexity this can be refactored this way:

```php
            $requestUri = $this->server->get('REQUEST_URI');

            if ('' !== $requestUri && '/' === $requestUri[0]) {
                // Remove the fragment.
                if (false !== $pos = strpos($requestUri, '#')) {
                    $requestUri = substr($requestUri, 0, $pos);
                }
            } else {
                // HTTP proxy reqs setup request URI with scheme and host [and port] + the URL path,
                // only use URL path.
                $uriComponents = parse_url($requestUri);

                if (isset($uriComponents['path'])) {
                    $requestUri = $uriComponents['path'];
                }

                if (isset($uriComponents['query'])) {
                    $requestUri .= '?'.$uriComponents['query'];
                }
            }
```